### PR TITLE
Add a system test for 'Table.upload_from_file'.

### LINKED ISSUE #2133: Calling 'job.begin' on job returned from 'Table.upload_from_file' raises a 400
Trying to load data into bigquery with `upload_from_file` returns a 400.

Sample code:

```python
def load_data_from_file(dataset_name, table_name, source_file_name):
    bigquery_client = bigquery.Client()
    dataset = bigquery_client.dataset(dataset_name)
    table = dataset.table(table_name)

    # Reload the table to get the schema.
    table.reload()

    with open(source_file_name, 'rb') as source_file:
        # This example uses CSV, but you can use other formats.
        # See https://cloud.google.com/bigquery/loading-data
        job = table.upload_from_file(
            source_file, source_format='text/csv')

    job.begin()

    wait_for_job(job)

    print('Loaded {} rows into {}:{}.'.format(
        job.output_rows, dataset_name, table_name))
```

Error:

```
Traceback (most recent call last):
  File "load_data_from_file.py", line 77, in <module>
    args.source_file_anme)
  File "load_data_from_file.py", line 45, in load_data_from_file
    job.begin()
  File "/Users/jonwayne/workspace/python-docs-samples/bigquery/cloud-client/env/lib/python3.4/site-packages/gcloud/bigquery/job.py", line 364, in begin
    method='POST', path=path, data=self._build_resource())
  File "/Users/jonwayne/workspace/python-docs-samples/bigquery/cloud-client/env/lib/python3.4/site-packages/gcloud/connection.py", line 347, in api_request
    error_info=method + ' ' + url)
gcloud.exceptions.BadRequest: 400 Load configuration must specify at least one source URI (POST https://www.googleapis.com/bigquery/v2/projects/python-docs-samples-tests/jobs)
```

### PR DESCRIPTION
**Pull request by tseaver**
Closes #2133.

/cc @jonparrott, @matheuspatury

I will cancel the PR job on Travis, as this PR changes only system tests.

### CODE REVIEW

**Comment by tseaver** [comment]
@jgeewax, @omaray the new system test passes, and shows correct usage of `Table.upload_from_file`.

**Comment by dhermes** [comment]
Before reviewing, this doesn't seem to fix the title of the issue:

> bigquery.table.Table.upload_from_file fails with 400

Should the issue be renamed to be more accurate?

File: system_tests/bigquery.py [hunk 1]
```diff
@@ -288,6 +288,58 @@ def _has_rows(result):
...
+        table.create()
+        self.to_delete.insert(0, table)
+
+        with tempfile.NamedTemporaryFile(mode='w+') as csv_file:
+            writer = csv.writer(csv_file)
+            writer.writerow(('Full Name', 'Age'))
+            writer.writerows(ROWS)
+            csv_file.flush()
+
+            with open(csv_file.name, 'rb') as csv_read:
```
Comment (dhermes) [inline system_tests/bigquery.py]: The read inside the write is a bit icky

File: system_tests/bigquery.py [hunk 2]
```diff
@@ -288,6 +288,58 @@ def _has_rows(result):
...
+
+        def _job_done(instance):
+            return instance.state.lower() == 'done'
+
+        # Retry until done.
+        retry = RetryInstanceState(_job_done, max_tries=8)
+        retry(job.reload)()
+
+        self.assertTrue(_job_done(job))
+        self.assertEqual(job.output_rows, len(ROWS))
```
Comment (dhermes) [inline system_tests/bigquery.py]: Why not assert values too?

File: system_tests/bigquery.py [hunk 3]
```diff
@@ -288,6 +288,58 @@ def _has_rows(result):
...
+                )
+
+        def _job_done(instance):
+            return instance.state.lower() == 'done'
+
+        # Retry until done.
+        retry = RetryInstanceState(_job_done, max_tries=8)
+        retry(job.reload)()
+
+        self.assertTrue(_job_done(job))
```
Comment (dhermes) [inline system_tests/bigquery.py]: Isn't this true by default since `retry` exited successfully?

File: system_tests/bigquery.py (hunk 1 above)
Comment (tseaver) [inline system_tests/bigquery.py]: If we drop that, the file won't exist to be read.

File: system_tests/bigquery.py (hunk 2 above)
Comment (tseaver) [inline system_tests/bigquery.py]: It is redundant with the other test, and can take longer than just checking the job stats.

File: system_tests/bigquery.py (hunk 3 above)
Comment (tseaver) [inline system_tests/bigquery.py]: Neither `RetryInstanceState` nor `RetryResult` raise exceptions if they "run off the end."

Comment (dhermes) [inline system_tests/bigquery.py]: Isn't that a bug?

File: system_tests/bigquery.py (hunk 2 above)
Comment (dhermes) [inline system_tests/bigquery.py]: Which other test do you refer to?

**Comment by tseaver** [comment]
> Before reviewing, this doesn't seem to fix the title of the issue: "bigquery.table.Table.upload_from_file fails with 400"

The OP in that issue made an understandable error, calling `job.begin()` on the job returned from `table.upload_from_file()`:  the issue is that the job *already* started, and the resource returned for the job does not contain enough information to (re)start it.  This PR just demonstrates that `table.upload_from_file()` works as expected without making that call. 

> Should the issue be renamed to be more accurate?

We could rename the issue, or just close it and move on.

**Comment by dhermes** [comment]
> We could rename the issue, or just close it and move on.

Can you rename it to be more accurate? This way if anyone comes across it again they won't be misled.

**Comment by tseaver** [comment]
Renamed.

File: system_tests/bigquery.py (hunk 2 above)
Comment (tseaver) [inline system_tests/bigquery.py]: `test_load_table_from_storage_then_dump_table`.

File: system_tests/bigquery.py (hunk 3 above)
Comment (tseaver) [inline system_tests/bigquery.py]: Maybe so, but it isn't related to this PR:  we routinely check the result after using them.

File: system_tests/bigquery.py (hunk 1 above)
Comment (daspecster) [inline system_tests/bigquery.py]: You could the `delete=False` option but then you would have to add delete logic at the end of the test or in a tearDown.

File: system_tests/bigquery.py (hunk 3 above)
Comment (dhermes) [inline system_tests/bigquery.py]: #2315 has been filed

File: system_tests/bigquery.py (hunk 2 above)
Comment (dhermes) [inline system_tests/bigquery.py]: If it's as easy as

```python
        by_age = operator.itemgetter(1)
        self.assertEqual(sorted(rows, key=by_age),
                         sorted(ROWS, key=by_age))
```

what's the incentive not to?

File: system_tests/bigquery.py (hunk 1 above)
Comment (dhermes) [inline system_tests/bigquery.py]: > If we drop that, the file won't exist to be read.

I realize, just wish there was a better way without having to wrap the entire method in a finally.

File: system_tests/bigquery.py (hunk 2 above)
Comment (tseaver) [inline system_tests/bigquery.py]: Done.

**Comment by dhermes** [comment]
LGTM

**Comment by tseaver** [comment]
@dhermes Any remaining issues?

**Pull request merged**